content/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-3.md

```diff
@@ -0,0 +1,295 @@
+
+![KafkaUtils createDirectStream method](/img/kafkautils-createdirectstream-method.png)
+
+![DStream](/img/dstream.png)
+
+In the code below each RDD in the valuesDStream is transformed into a Spark Dataset. Then the MapR Database Spark Connector DStream saveToMapRDB method performs a parallel partitioned bulk insert of JSON FlightwPred objects into MapR Database.
+
+![RDD in the valuesDStream](/img/rdd-in-valuesdstream.png)
+
+![Save DStream to MapR Database JSON](/img/save-dsstream-mapr-db-json.png)
+
+**Querying MapR Database JSON with Spark SQL**
+
+The Spark MapR Database Connector enables users to perform complex SQL queries and updates on top of MapR Database using a Spark Dataset, while applying critical techniques such as Projection and filter pushdown, custom partitioning, and data locality.
+
+![Application](/img/application.png)
+
+A Spark Dataset is a distributed collection of data. Dataset is a newer interface, which provides the benefits of strong typing, the ability to use powerful lambda functions, efficient object serialization/deserialization, combined with the benefits of Spark SQL's optimized execution engine.
+
+![Spark Dataset](/img/spark-dataset.png)
+
+A DataFrame is a Dataset organized into named columns Dataset\[Row]. (In Spark 2.0, the DataFrame APIs merged with Datasets APIs.)
+
+![Unified Apache Spark 2.0 API](/img/unified-apache-spark.png)
+
+**Loading data from MapR Database into a Spark Dataset**
+
+To [load data from a MapR Database JSON](https://docs.datafabric.hpe.com/62/Spark/LoadDataFromMapRDBasDataset.html) table into an Apache Spark Dataset we invoke the loadFromMapRDB method on a SparkSession object, providing the tableName, schema and case class. This will return a Dataset of FlightwPred objects:
+
+![load data from a MapR Database JSON](/img/load-data-from-mapr-db-json.png)
+
+**Explore and query the Flight data with Spark SQL**
+
+Datasets provide a domain-specific language for structured data manipulation in Scala, Java, and Python. Below are some examples in scala. The Dataset show() action displays the top 20 rows in a tabular form.
+
+![top 20 rows in a tabular form](/img/top-20-rows-tabular-form.png)
+
+In the code below a filter is used to count the predicted delays, actual delays and total. This is then used to calculate the ratio wrong, correct, false positive. These type of calculations would be useful for continued analysis of models in production.
+
+![delays, actual delays and total](/img/filter-count-predicted-delays.png)
+
+The output is shown below.
+
+![Output of predicted delays, actual delays and total](/img/delays-actual-delays-total-output.png)
+
+**What is the count of predicted delay/notdelay for this dstream dataset?**
+
+![What is the count of predicted delay/notdelay for this dstream dataset?](/img/count-predicted-dstream-dataset.png)
```

Comment:
This is another image that really looks kind of fuzzy given how large it is. Several others follow.

![image](https://user-images.githubusercontent.com/77496261/137183782-29faa072-46d2-45ef-bbd7-707aa8b0b1c4.png)

Reply:
done